src/controller/CHIPDeviceController.cpp

```diff
@@ -894,8 +894,34 @@ CHIP_ERROR DeviceCommissioner::StopPairing(NodeId remoteDeviceId)
 
 CHIP_ERROR DeviceCommissioner::UnpairDevice(NodeId remoteDeviceId)
 {
-    // TODO: Send unpairing message to the remote device.
-    return CHIP_NO_ERROR;
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    MATTER_TRACE_EVENT_SCOPE("UnpairDevice", "DeviceCommissioner");
+
+    Platform::UniquePtr<DeviceUnpair> deviceUnpair(new DeviceUnpair());
+    deviceUnpair->RegisterCallback(this);
+    CommissioneeDeviceProxy * commissioneeDeviceProxy = nullptr;
+    if (CHIP_NO_ERROR == GetDeviceBeingCommissioned(remoteDeviceId, &commissioneeDeviceProxy))
+    {
+        deviceUnpair->UnpairDevice(commissioneeDeviceProxy, remoteDeviceId);
+    }
+    else
+    {
+        err = GetConnectedDevice(remoteDeviceId, &deviceUnpair->GetConnectedCallback(),
+                                 &deviceUnpair->GetConnectionFailureCallback());
+    }
+
+    mDeviceUnpair = std::move(deviceUnpair);
```

Comment:
So if I call `UnpairDevice` a second time before the first call is done, we get use after free on the first UnpairDevice object, right?

And there's no way to know when the first call is or is not done....

Reply:
I added exception code. (check mDeviceUnpair null or not.)